FIX Enable setting of sub-parameters for deprecated base_estimator param

betatim opened the pull request:
#### Reference Issues/PRs
closes #25470


#### What does this implement/fix? Explain your changes.
The "base_estimator" param was renamed to "estimator" which lead to users not being able to set parameters for "base_estimator__*" properties via `set_params` anymore. For example when using the old names to define the parameter grid used by `RandomSearchCV` and friends.

This introduces a special translation in `BaseEstimator.set_params`. When the key is "base_estimator" it gets changed to "estimator". This mirrors the "translation" performed in `BaseEnsemble`.


#### Any other comments?
This is an interesting problem. The pattern of setting a parameter that has been deprecated to `"deprecated"` is used a lot, but probably the parameter isn't normally a "thing that has parameters itself". Another tricky thing is that `BaseEstimator.set_params` has no way of knowing what the new name of the parameter is (that `base_estimator` became `estimator`). This means it is tricky to write a generic "translation" layer.

However, after a quick search I think `base_estimator` is the only deprecated parameter that has "sub parameters". This would explain why despite wide usage of the pattern, this bug hasn't come up before. This is also why I put this in `BaseEstimator`.

I think there is a bit of tension here. To support the deprecation cycle you should allow users to do:
```python
clf = AdaBoostClassifier(DecisionTreeClassifier())
clf.set_params(base_estimator__max_depth=42)
```
because that is what their existing code probably looks like. However, this makes `clf.get_params()` a bit weird. It will show that `estimator__max_depth` was set to 42, and `base_estimator__max_depth` doesn't feature at all. This feels a bit weird.

Another weird thing that is possible after this PR is the following:
```python
clf = AdaBoostClassifier(estimator=DecisionTreeClassifier())
clf.set_params(base_estimator__max_depth=42)
```
I'm not sure we can detect this case though. It feels like this should lead to an exception, but unsure how to detect this case and if it is worth the effort.

betatim committed:
Enable setting of sub-parameters for base_estimator param

The base_estimator param was renamed to estimator which lead to users
not being able to set parameters for base_estimator__* properties.

glemaitre commented on sklearn/ensemble/tests/test_weight_boosting.py:
```diff
@@ -630,3 +630,20 @@ def test_base_estimator_property_deprecated(AdaBoost):
+    # "base_estimator".
+    clf = AdaBoostClassifier(DecisionTreeClassifier())
+
+    clf.set_params(base_estimator__max_depth=2)
```
We would expect the following to succeed.

```suggestion
    with pytest.warns(FutureWarning):
        clf.set_params(base_estimator__max_depth=2).fit(X, y)
```

glemaitre commented on sklearn/base.py:
```diff
@@ -227,6 +227,12 @@ def set_params(self, **params):
+            # renamed to "estimator". This means we need to translate it
+            # here and set sub-parameters on "estimator", but only if the
+            # user did not explicitly set a value for "base_estimator".
+            if key == "base_estimator" and valid_params[key] == "deprecated":
+                key = "estimator"
```
This is really tricky here:

- `base_estimator` should not be overwritten by scikit-learn
- However, we need to indicate to our users that they use a deprecated attribute that is triggered usually by modifying `base_estimator`. So the natural way would be to modify `base_estimator`

I have to think more about the right workaround.

glemaitre replied:
If we don't care about the first rule (mutate the original attribute):

```python
            if key == "base_estimator" and valid_params[key] == "deprecated":
                self.set_params(base_estimator=clone(valid_params["estimator"]))
                self.set_params(estimator=None)
                valid_params["base_estimator"] = self.base_estimator
```

This would set the estimator to raise the warning and set the parameters.

ogrisel replied:
Why wouldn't we just raise the warning in `set_params` directly instead, something like:

```python
            if key == "base_estimator" and valid_params[key] == "deprecated":
                warnings.warn(
                    f"Parameter 'base_estimator' of {self.__class__} is deprecated "
                    f"in favor of 'estimator'. See {self.__class__}'s docstring "
                    "for more details"
                )
                key = "estimator"
                valid_params[key].set_params(**sub_params)
```

ogrisel replied:
Maybe we could use a `stacklevel=2` on top of that but I am never sure about which value to use for `stacklevel` :)

glemaitre replied:
Yep, it looks simpler and it will raise fewer warnings (1 during `set_params` instead of 1 at each `fit` call). Maybe, we can restrict ourselves to scikit-learn module looking at the module of `self` to not impact 3rd party even thought the pattern `"deprecated"` indicate that 3rd party want this change.

betatim replied:
Implemented the warning idea from Olivier together with the restriction to estimators that are part of scikit-learn.

betatim committed:
Raise a warning when using old parameter names

betatim committed:
Add what's new entries

The list of effected estimators comes from looking at the v1.2 what's
new to find where the renaming took place.

betatim commented:
I made two what's new entries, one for each module that changed from `base_estimator` to `estimator`. I found the list of estimators that were changed by looking at the v1.2 what's new.

glemaitre commented on sklearn/base.py:
```diff
@@ -227,6 +227,24 @@ def set_params(self, **params):
                 valid_params[key] = value
 
         for key, sub_params in nested_params.items():
+            # The "base_estimator" key is special. It was deprecated and
+            # renamed to "estimator" for several estimators. This means we
+            # need to translate it here and set sub-parameters on "estimator",
+            # but only if the user did not explicitly set a value for
+            # "base_estimator".
```
```suggestion
            # TODO(remove 1.4)
            # The "base_estimator" key is special. It was deprecated and
            # renamed to "estimator" for several estimators. This means we
            # need to translate it here and set sub-parameters on "estimator",
            # but only if the user did not explicitly set a value for
            # "base_estimator".
```

glemaitre commented on sklearn/base.py:
```diff
@@ -227,6 +227,24 @@ def set_params(self, **params):
                 valid_params[key] = value
 
         for key, sub_params in nested_params.items():
+            # The "base_estimator" key is special. It was deprecated and
```
```suggestion
            # TODO(1.4): remove specific handling of "base_estimator".
            # The "base_estimator" key is special. It was deprecated and
```

glemaitre commented on sklearn/ensemble/tests/test_weight_boosting.py:
```diff
@@ -630,3 +630,21 @@ def test_base_estimator_property_deprecated(AdaBoost):
+    # This implicitly sets "estimator", it is how old code (pre v1.2) would
+    # have instantiated AdaBoostClassifier and back then it would set
+    # "base_estimator".
+    clf = AdaBoostClassifier(DecisionTreeClassifier())
```
actually here we are setting `estimator` and not `base_estimator`. It is completely fine that it works. However, could we parametrize the non-regression test to check that both works when setting `estimator` and `base_estimator`.

It could also be work to call `fit` just to make sure that there is not latter side-effect.

ogrisel replied:
I think this test is enough. If `clf = AdaBoostClassifier(base_estimator=DecisionTreeClassifier())` then I would expect that the warning would be raised at `__init__` time and no warning would be raised when calling `clf.set_params(base_estimator__max_depth=2)`.

betatim replied:
This is indeed what happens. I also think the warning at `fit` time is enough in this case.

glemaitre commented on doc/whats_new/v1.3.rst:
```diff
@@ -96,6 +96,13 @@ Changelog
 - |Feature| A `__sklearn_clone__` protocol is now available to override the
   default behavior of :func:`base.clone`. :pr:`24568` by `Thomas Fan`_.
 
+:mod:`sklearn.calibration`
```
I would move this in 1.2.2 because this is an unfortunate regression.

betatim replied:
Done

glemaitre commented on doc/whats_new/v1.3.rst:
```diff
@@ -136,6 +143,12 @@ Changelog
   scikit-learn 1.3: retraining with scikit-learn 1.3 is required.
   :pr:`25186` by :user:`Felipe Breve Siola <fsiola>`.
 
+- |Fix| A deprecation warning is raised when using the "base_estimator__" prefix
```
Then, I think that we don't need this entry. We can alterate the previous 1.2 entry mentioning that the warning is also announce when calling `set_params`.

ogrisel replied:
I think this need to me moved to a dedicated entry targeting the 1.2.2 version.

betatim replied:
I've moved the change log entries to the 1.2.2 what's new. So we now have a dedicated entry.

glemaitre reviewed (commented):
Just some minor comment. I added the 1.2.2 milestone since this is a regression .

betatim committed:
Merge branch 'main' into fix-25470

betatim committed:
Move what's new entries, add TODO marker

betatim committed:
Merge branch 'main' into fix-25470

betatim committed:
Fix missing newline

betatim committed:
Fix what's new syntax

thomasjpfan commented on sklearn/base.py:
```diff
@@ -227,6 +227,25 @@ def set_params(self, **params):
+                and self.__module__.startswith("sklearn.")
+            ):
+                warnings.warn(
+                    f"Parameter 'base_estimator' of {self.__class__.__name__} is"
```
**Not actionable**: It would be harder for third party estimators to deprecate a positional parameter because they would also need adjust `set_params` to still work with the previous name.

thomasjpfan reviewed (approved):
Thanks for the fix! I think this is a good enough fix for scikit-learn estimators. LGTM

jeremiedbb committed:
Merge branch 'main' into fix-25470

jeremiedbb reviewed (approved):
Since we check the module to be sklearn, this PR should only impact our estimators, so I'm fine with this temporary solution.

LGTM
